File: src/blocks/Brickwall.php
<?php

namespace jmw\fabricad\blocks;

use jmw\fabricad\shapes\Shape;
use jmw\fabricad\shapes\Container;
use jmw\fabricad\shapes\BinaryOperators;
use jmw\fabricad\shapes\Polygon;
use jmw\fabricad\shapes\Line;
use jmw\fabricad\shapes\Point;
use jmw\fabricad\shapes\Rectangle;

class Brickwall extends BasicBuildingBlock
{
    
    public function getStart(): int
    {
        if (isset($this->config['start'])) {
            return $this->config['start'];
        } else {
            return 0;
        }
    }
    
    public function getBrickHeight(): float {
        if (isset($this->config['height'])) {
            return $this->config['height'];
        } else {
            return 5.0;
        }
    }
    
    public function getBrickWidth(): float {
        if (isset($this->config['width'])) {
            return $this->config['width'];
        } else {
            return 10.0;
        }
    }
    
    public function isHorizontal(): bool
    {
        if (isset($this->config['vertical'])) {
            return ($this->config['vertical'] > 0);
        } else {
            return true;
        }
    }
    
    /**
     * 
     * @param Rectangle $bb
     * @return array
     */
    protected function renderBricks(Rectangle $bb): array
    {
        $totalheight = $bb->getTop()->getY() + $this->getBrickHeight();
        $prevY = $bb->getOrigin()->getY();
        $startX = $bb->getOrigin()->getX();
        $endX = $bb->getTop()->getX() + $this->getBrickWidth();
        
        $lines = array();
        
        $counter = $this->getStart();
        for($y = $bb->getOrigin()->getY() ; $y <= $totalheight ; $y += $this->getBrickHeight() ) {
            $lines['horizontal'][] = new Line(new Point($startX, $y), new Point($endX, $y ));
            $start = $startX + (1-($counter % 2)/2) * $this->getBrickWidth();
            for($x = $start ; $x < $endX ; $x += $this->getBrickWidth() ) {
                $lines['vertical'][] = new Line(new Point($x, $prevY), new Point($x, $y));
            }
            $prevY = $y;
            $counter++;
        }
        
        return $lines;
    }
    
    public function render(): Shape
    {
        $bb = $this->getShape()->getBoundingBox();
        
        if ($this->isHorizontal()) {
            
            $bricks = $this->renderBricks($bb);
            
            $lines = array_merge($bricks['vertical'], $bricks['horizontal']);
            
        } 
        
        $c = new Container();
        //foreach($this->getShape()->asPolygon()->getLines() as $line) {
        //    $c->addShape($line);
        //}
        if ($this->getShape() instanceof Polygon) {
            foreach($lines as $l) {
                // echo "I have: ".$l."\n";
                
                $items = BinaryOperators::intersection($this->getShape()->asPolygon(), $l);
                
                // echo "I get: \n";

                foreach($items as $it) {
                    // echo " - ".$it;
                    $c->addShape($it);
                }
            }
        }
        
        $c->addShape($this->getShape());
        
        return $c;
    }
}

File: src/blocks/Brickwall.cc


#include "Brickwall.h"


namespace fabricad::blocks
{
  Brickwall::Brickwall()
  {
    brick_height_ = 5.0;
    brick_width_ = 10.0;
    start_ = 0;
  }

  float Brickwall::getBrickHeight()
  {
    return brick_height_;
  }

  Brickwall* Brickwall::setBrickHeight(float height)
  {
    brick_height_ = height;
    return this;
  }

  float Brickwall::getBrickWidth()
  {
    return brick_width_;
  }

  Brickwall* Brickwall::setBrickWidth(float width)
  {
    brick_width_ = width;
    return this;
  }

  size_t Brickwall::getStartRow()
  {
    return start_;
  }

  Brickwall* Brickwall::setStartRow(size_t start)
  {
    start_ = start;
    return this;
  }

  std::string Brickwall::toString(std::string indent)
  {
    std::string s = BasicBuildingBlock::toString(indent);
    s = s + indent + "Bricks (w x h): " + std::to_string(getBrickWidth()) + ", " + std::to_string(getBrickHeight()) + "\n";
    s = s + indent + "Start at row  : " + std::to_string(getStartRow()) + "\n";

    return s;
  }

  void Brickwall::render()
  {
    // Create initial layer with shape
    BasicBuildingBlock::render();
    // Now, create second layer with bricks

    box bb;
    bg::envelope(this->shape_, bb);
    std::vector<linestring> lines;

    renderBricksFor(bb, lines);

    for(size_t i = 0 ; i < lines.size() ; i++)
    {
      std::vector<linestring> l;

      bg::intersection(lines[i], this->shape_, l);

      this->layers_[1].lines.insert(
            this->layers_[1].lines.end(),
            l.begin(),
            l.end()
      );
    }
  }

  void Brickwall::renderBricksFor(box const& bb, std::vector<linestring> &lines)
  {
    float minX = bb.min_corner().get<0>();
    float minY = bb.min_corner().get<1>();
    float maxX = bb.max_corner().get<0>() + getBrickWidth();
    float maxY = bb.max_corner().get<1>() + getBrickHeight();

    size_t counter = getStartRow();
    float prevY = minY;
    // Generate horizontal lines
    // Notice that we use <=, as we still want to draw vertical lines!
    for(float i = minY + getBrickHeight() ; i <= maxY ; i += getBrickHeight() )
    {
      point p1 = point(minX, i);
      point p2 = point(maxX, i);
      if (i < maxY) {
        // If it is  >=, we do not need to draw it, as it will fall outside of
        // the shape anyway.
        lines.push_back(linestring({p1, p2}));
      }

      float start = minX + (1-((float) (counter % 2))/2) * getBrickWidth();
      for(float j = start ; j <= maxX ; j += getBrickWidth() )
      {
        point py1 = point(j, prevY);
        point py2 = point(j, i);
        lines.push_back(linestring({py1, py2}));
      }
      prevY = i;
      counter++;
    }
  }
}
